.github/workflows/ci.yml

```diff
@@ -0,0 +1,60 @@
+name: CI
+
+on:
+  push:
+    # Avoid duplicate builds on PRs.
+    branches:
+      - main
+    tags: 
+      - v*
+  pull_request:
+
+permissions:
+  contents: read
+
+jobs:
+  shellcheck:
+    runs-on: ubuntu-22.04
+    container:
+      image: koalaman/shellcheck-alpine:v0.8.0
+    steps:
+    - uses: actions/checkout@v2
+    - name: Run shellcheck
+      run: find . -type f \( -name "*.sh" -o -path "*/bin/*" \) ! -name '*.jq' | xargs -t shellcheck
+  build-stack:
+    runs-on: ubuntu-22.04
+    needs:
+    - shellcheck
+    env:
+      STACK: heroku-${{ matrix.stack-version }}
+      STACK_VERSION: "${{ matrix.stack-version }}"
+      DOCKER_HUB_TOKEN: "${{ secrets.DOCKER_HUB_TOKEN }}"
+      ID_SERVICE_PASSWORD: "${{ secrets.ID_SERVICE_PASSWORD }}"
+      MANIFEST_APP_TOKEN: "${{ secrets.MANIFEST_APP_TOKEN }}"
+    strategy:
+      matrix:
+        stack-version:
+        - '18'
+        - '20'
+        - '22'
+    steps:
+    - uses: actions/checkout@v2
+    - name: Build images
+      run: bin/build.sh $STACK_VERSION
+    - name: Check that the generated files are in sync
+      run: |-
+        status="$(git status --porcelain)"
+        if [[ -n "$status" ]]; then
+          echo "Generated files differ from checked-in versions! Run bin/build.sh to regenerate them locally."
+          echo -e "\nChanged files:\n${status}\n"
+          git diff
+          exit 1
+        fi
+    - name: Publish to image registries
+      run: bin/publish-to-registries.sh
+      if: success() && (github.ref_name == 'main' || github.ref_type == 'tag')
+    - name: Convert docker image and release to Heroku staging
+      run: bin/convert-and-publish-to-heroku.sh
+      if: success() && github.ref_type == 'tag'
+      env:
+        GITHUB_TAG: ${{ github.ref_name }}
```

Comment:
(optional)

Initially I thought there was a bug here, since this step passes `GITHUB_TAG`, but yet the step before (which also needs to run on tags) does not pass `GITHUB_TAG` (ie it seems the step prior should too).

However, `bin/publish-to-registries.sh` uses a different approach - it has a `"$GITHUB_REF_TYPE" == 'tag'` conditional, and then uses `GITHUB_REF_NAME` (which is equal to the tag name, so long as the first conditional is true).

I wonder if having these two different approaches might be confusing to future maintainers? In which case, maybe it's worth either:
1. Switching both to `GITHUB_TAG` (though that would need to not set the env var when `github.ref_name == 'main'` is the case),
2. Or else, removing the concept of `GITHUB_TAG` and adding `GITHUB_REF_TYPE" == 'tag'` +  `GITHUB_REF_NAME` to `tools/bin/capture-docker-stack`?

Reply:
yeah, so the previous logic in `bin/publish-to-registries` was conditional on the `CIRCLE_TAG` which is only set if the ref type is a tag. Not knowing if / how / when that script will be run, and in the interest of limiting setting custom env variables I thought using the ref_type was the correct flag.

I don't think it will be any more confusing for future maintainers than it was, in fact I think it might be less confusing because it's a more explicit check on the ref_type.

So I'd prefer option #2 here and update the logic in `capture-docker-stack` to check the same. Let me do that.